inc/pkthlp.h

```diff
@@ -309,6 +309,158 @@ PktBuildTcpFrame(
+        *payloadptr = Payload[i];
+        payloadptr = payloadptr + 1;
+    }
+    Icmp4Hdr->Checksum = PktChecksum(0, Icmp4Hdr, sizeof(*Icmp4Hdr) + PayloadLength);
+
+    *BufferSize = TotalLength;
+
+    return TRUE;
+}
+
+inline
+_Success_(return != FALSE)
+BOOLEAN
+PktBuildIcmp6Frame(
+    _Out_ VOID *Buffer,
+    _Inout_ UINT32 *BufferSize,
+    _In_ CONST UCHAR *Payload,
+    _In_ UINT16 PayloadLength,
+    _In_ CONST ETHERNET_ADDRESS *EthernetDestination,
+    _In_ CONST ETHERNET_ADDRESS *EthernetSource,
+    _In_ CONST VOID *IpDestination,
+    _In_ CONST VOID *IpSource
+    )
+{
+    CONST UINT32 TotalLength = sizeof(ETHERNET_HEADER) + sizeof(IPV6_HEADER) + sizeof(ICMPV6_HEADER) + PayloadLength;
+    if (*BufferSize < TotalLength) {
+        return FALSE;
+    }
+
+    void* PktBuffer = Buffer;
+
+    // Fill Ethernet headers.
+    ETHERNET_HEADER *EthernetHeader = (ETHERNET_HEADER *) PktBuffer;
+    EthernetHeader->Destination = *EthernetDestination;
+    EthernetHeader->Source = *EthernetSource;
+    EthernetHeader->Type = htons(ETHERNET_TYPE_IPV6);
+    PktBuffer = EthernetHeader + 1;
+
+    // Fill IP and ICMP headers
+    IPV6_HEADER *IpHeader = (IPV6_HEADER *)PktBuffer;
+    IpHeader->VersionClassFlow = IPV6_VERSION;
+    IpHeader->PayloadLength = htons(sizeof(ICMPV6_HEADER) + PayloadLength);
+    IpHeader->NextHeader = IPPROTO_ICMPV6;
+    IpHeader->HopLimit = 1;
+    UINT8 AddressLength = sizeof(IN6_ADDR);
+    RtlCopyMemory(&IpHeader->SourceAddress, IpSource, AddressLength);
+    RtlCopyMemory(&IpHeader->DestinationAddress, IpDestination, AddressLength);
+    PktBuffer = IpHeader + 1;
+    ICMPV6_HEADER *Icmp6Hdr = (ICMPV6_HEADER *) PktBuffer;
+    Icmp6Hdr->Type = 129;
+    Icmp6Hdr->Code = 0;
+    PktBuffer = Icmp6Hdr + 1;
+
+    // Fill ICMP payload
+    UCHAR *payloadptr = (UCHAR *)PktBuffer;
+    for (int i = 0; i < PayloadLength; i++) {
+        *payloadptr = Payload[i];
+        payloadptr = payloadptr + 1;
+    }
+
+    //
+    // Prepare data to compute ICMPV6 checksum
+    // Data for checksum calculation: { IPV6 pseudo header + ICMPv6 header (0 for cxsum) + ICMPv6 payload }
+    //
+    Icmp6Hdr->Checksum = 0;
+    UINT16 CheckSumDataLength = 2 * sizeof(IN6_ADDR) + sizeof(UINT32) + sizeof(UINT32) + sizeof(ICMPV6_HEADER) + PayloadLength;
+    void *ptr = malloc(CheckSumDataLength);
+    if (ptr == NULL) {
+        return FALSE;
+    }
+    memset(ptr, 0, CheckSumDataLength);
+    UCHAR *CheckSumData = (UCHAR*) ptr;
+    IN6_ADDR *src = (IN6_ADDR*) ptr;
+    *src = IpHeader->SourceAddress;
+    ptr = src + 1;
+    IN6_ADDR *dst = (IN6_ADDR*) ptr;
+    *dst = IpHeader->DestinationAddress;
+    ptr = dst + 1;
+    UINT16 UpperLayerPacketLen = sizeof(ICMPV6_HEADER) + PayloadLength;
+    UINT32 *upperlayerpacketlen = (UINT32 *) ptr;
+    *upperlayerpacketlen = htons(UpperLayerPacketLen);
+    ptr = upperlayerpacketlen + 1;
+    UCHAR* reserved = (UCHAR*) ptr;
+    // zeroes for reserved bits
+    ptr = reserved + 3;
```

Comment:
I'm not sure why there's so much pointer manipulation happening here - can't we use/define a struct?

Reply:
Fixed in the next iteration.